toolbox/deps_update/main.go

```diff
@@ -0,0 +1,201 @@
+// useful in avoiding making duplicate branches of the same code change
+func fingerPrint(deps []dependency) (string, error) {
+	digest := ""
+	for _, dep := range deps {
+		commitSHA, err := githubClnt.getHeadCommitSHA(dep.RepoName, dep.ProdBranch)
+		if err != nil {
+			return "", err
+		}
+		digest = digest + commitSHA
+	}
+	return GetMD5Hash(digest), nil
+}
+
+// Update the commit SHA reference in the dependency file of dep
+func updateDepFile(dep dependency) error {
+	input, err := ioutil.ReadFile(dep.File)
+	if err != nil {
+		return err
+	}
+	lines := strings.Split(string(input), "\n")
+	for i, line := range lines {
+		if strings.Contains(line, dep.Name+" = ") {
+			if lines[i], err = replaceCommit(line, dep); err != nil {
+				return err
+			}
+		}
+	}
+	output := strings.Join(lines, "\n")
+	return ioutil.WriteFile(dep.File, []byte(output), 0600)
+}
+
+// Assumes at the root of istio directory
+// Runs the updateVersion.sh script
+func updateIstioDeps() error {
+	pilotSHA, err := githubClnt.getHeadCommitSHA("pilot", "stable")
+	if err != nil {
+		return err
+	}
+	mixerSHA, err := githubClnt.getHeadCommitSHA("mixer", "stable")
+	if err != nil {
+		return err
+	}
+	istioctlURL := fmt.Sprintf("https://storage.googleapis.com/istio-artifacts/pilot/%s/artifacts/istioctl", pilotSHA)
+	hub := "gcr.io/istio-testing"
+	cmd := fmt.Sprintf("./install/updateVersion.sh -p %s,%s -x %s,%s -i %s", hub, pilotSHA, hub, mixerSHA, istioctlURL)
+	_, err = Shell(cmd)
+	return err
+}
+
+// Update the given repository so that it uses the latest dependency references
+// push new branch to remote, create pull request on master,
+// which is auto-merged after presumbit
+func updateDeps(repo string) error {
+	if _, err := Shell("rm -rf " + repo); err != nil {
+		return err
+	}
+	remoteURL, err := remote(repo)
+	if err != nil {
+		return err
+	}
+	if _, err = Shell("git clone " + remoteURL); err != nil {
+		return err
+	}
+	if err = os.Chdir(repo); err != nil {
+		return err
+	}
+	deps, err := getDeps()
+	if err != nil {
+		return err
+	}
+	depVersions, err := fingerPrint(deps)
+	if err != nil {
+		return err
+	}
+	branch := "autoUpdateDeps" + depVersions
+	if _, err := Shell("git checkout -b " + branch); err != nil {
+		return err
+	}
+	if repo == "istio" {
+		if err := updateIstioDeps(); err != nil {
+			return err
+		}
+	} else {
+		for _, dep := range deps {
+			if err := updateDepFile(dep); err != nil {
+				return err
+			}
+		}
+	}
+	if _, err := Shell("git add *"); err != nil {
+		return err
+	}
+	if _, err := Shell("git commit -m Update_Dependencies"); err != nil {
+		return err
+	}
+	if _, err := Shell("git push --set-upstream origin " + branch); err != nil {
+		return err
+	}
+	if err := githubClnt.createPullRequest(branch, repo); err != nil {
+		return err
+	}
+	if err := os.Chdir(".."); err != nil {
+		return err
+	}
+	if _, err := Shell("rm -rf " + repo); err != nil {
+		return err
+	}
+	return nil
+}
+
+func main() {
```

Comment:
yeah so another option is for handling the contexts is to have ctx := context.Background() here and then thread that through all the network calls.